src/celeritas/ext/detail/TrackProcessor.hh

```diff
@@ -0,0 +1,92 @@
+//------------------------------- -*- C++ -*- -------------------------------//
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/ext/detail/TrackProcessor.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "corecel/Macros.hh"
+#include "celeritas/Types.hh"
+
+class G4ParticleDefinition;
+class G4Step;
+class G4Track;
+class G4VProcess;
+class G4VUserTrackInformation;
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Manage track information for reconstruction.
+ *
+ * This class handles the bookkeeping of Geant4 track information needed
+ * to reconstruct tracks during hit processing. It maintains mappings between
+ * Celeritas PrimaryID and Geant4 track data.
+ *
+ */
+class TrackProcessor
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using VecParticle = std::vector<G4ParticleDefinition const*>;
+    //!@}
+
+  public:
+    // Construct with particle definitions for track reconstruction
+    explicit TrackProcessor(VecParticle const&);
+
+    ~TrackProcessor();
+    CELER_DEFAULT_MOVE_DELETE_COPY(TrackProcessor);
+
+    // Clear G4Track reconstruction data
+    void end_event();
+
+    // Register mapping from Celeritas PrimaryID to Geant4 TrackID
+    [[nodiscard]] PrimaryId register_primary(G4Track&);
+
+    // Restore track information for given primary and particle IDs
+    G4Track& restore_track(ParticleId, PrimaryId) const;
```

Comment:
Not sure if we want to have this be a const function. It modifies the temporary step/tracks so it's not const but it's called from const functions in the HitProcessor